File: Week-01/Day-01/Ques/main.cpp
#include <iostream>
using namespace  std;



// Traversal and Print: 
void PrintArr(int arr[], int size){
  for (int i = 0; i < size; i++){
    cout << arr[i] << " ";
  }
  cout << endl;
}

// Insert an Element at a Specific Position:
void InsertAtPosition(int arr[], int &size, int position, int element,int capacity) { // &size = passed by reference (means value reflect outside of the function)
  if (position < 0 || position > size || size >= capacity){ 
    cout << "Invalid position or array is full" << endl;
    return;
  }
  // shift elements to the right from the end of the position
  for (int i = size; i > position; i--) {
    arr[i] = arr[i-1];
  }

  // Insert the new element 
  arr[position] = element;
  size++;
}

// Delete an Element from a Specific Position:
void DeleteAtPosition (int arr[],int &size, int position, int element, int capacity) {
  if (position < 0 || position > size || size >= capacity) {
    cout << "Invalid position or array is full" << endl;
    return;
  }

  for (int i = size; i > position-1; i++){
arr[position] = arr[position+1];
  }
}


// Find Maximum and Minimum Element in an Array: 
void minMaxArr(int  arr[], int &size){
  int min = arr[0];
  // For min: 
  for (int i = 0; i < size; i++){
    if (min > arr[i]){
      min = arr[i];
    }
  }
  // For max:
  int max = arr[0];
  for (int i = 0; i < size; i++){
    if (max < arr[i]){
      max = arr[i];
    }
  }
  cout << "Min element in Array: "<<  min << "\n";
  cout << "Max element in Array: "<<  max << "\n";
}

// Rotate an Array (Left or Right):

// [1, 2, 3, 4, 5]
// [3, 4, 5, 1, 2]
void leftRotate(int arr[], int n, int k){
  
  // To handle cases where k is greater than n 
  k = k%n;


  // Temporary array stored elements to be rotated
  int temp[k]; 

  // Copy first k element to the temp array:
  for (int i = 0; i < k; i++){
    temp[i] = arr[i];
  }

  //shift remaining elements of arr[] to the left 
  for (int i = k; i < n; i++){
    arr[i-k] = arr[i];
  }

  // Copy elements from temp[] back to arr[] to complete rotation
  for (int i = 0; i < k; i++){
    arr[n-k+i] = temp[i];
  }
}



// [1, 2, 3, 4, 5]
// [4, 5, 1, 2, 3]
void rightRotate(int arr[], int n, int k){
  k = k%n;

  int temp[k];

  for (int i = n-k; i < n ;i++){
    temp[i-(n-k)] = arr[i];
 }
  for(int i= n-1; i >= k; i--){
    arr[i] = arr[i-k];
  }

  for (int i = 0; i < k;i++){
    arr[i] = temp[i];
  }
}

// Search in a Rotated Sorted Array:
int Search(int arr[], int &size, int element){
  for (int i = 0; i < size; i++){
    if (arr[i] == element){
      return 0; 
    } 
  }
  return -1;
}

// Delete from specific:
void Delete(int arr[], int &size, int index){
  for (int i = index; i < size-1; i++){
    arr[i] = arr[i+1];
  }
  size--;
}




int main(){
  cout << "Day -1 Array Questions: \n";

  // Traversal and Print: 
  int intArr[6] = {1,2,3,4,5};

  int size = 5;

  PrintArr(intArr,size);

  // Insert an Element at a Specific Position:
  int position = 3;
  int element = 71;
  int capacity = 6;
  InsertAtPosition(intArr, size, position, element,capacity);
  PrintArr(intArr, size);
  cout << "size increated: "<< size << "\n";

  // min max:

  int arr[10] = {71,69,22,7,33};
  size = 5;
  minMaxArr(arr,size);

  // Rotation:
  cout << "Original Array: ";
  PrintArr(arr,size);

  int n = 5; 
  int k = 2;
  // Left Rotation:
  cout << "Left Rotated Array: ";
  leftRotate(arr, n, k);
  PrintArr(arr,size);

  //  Right Rotation:

  cout << "Original Array: ";
  PrintArr(arr,size);
  cout << "Right Rotation: ";
  rightRotate(arr,n,k);
  PrintArr(arr, size);

  // Search in a Rotated Sorted Array:
  element = 7;
  cout <<"Is element found in the Array: " << Search(arr, size, element);
}

File: Week-01/Day-01/Ques/main.go
package main

import (
	"fmt"
)

func main() {
	// array initialization:
	// arr := [5]int{1, 2, 3, 4, 5} // this is array initliazation
	arr := []int{22, 71, 69, 7, 33, 1} // this is called slice in Golang without specifying size. Dynamic size

	fmt.Println("Array: ", arr)

	// Array Length:
	fmt.Println("Length of the Array: ", lenArr(arr))

	// Traversing array:
	traverse(arr)

	// Inserting an Element:
	index := 2
	element := 22
	fmt.Printf("Inserting %d in the index %d : ", element, index)
	fmt.Println(insArr(arr, index, element))

	// Delete Element:
	idx := 0
	fmt.Printf("Deleting element in the index %d : ", idx)

	DeleteElement := delArr(arr, idx)
	fmt.Println(DeleteElement)

	// Update Element:

	indx := 1
	new := 10
	fmt.Print("Updated array: ")
	fmt.Println(updateArr(arr, indx, new))

	// Access Element:
	key := 71
	result := AccessElement(arr, key)

	if result != -1 {
		fmt.Printf("%d is  found in the Array\n", key)
	} else {
		fmt.Println("Key not found in the Array :)")
	}

	// MAX
	fmt.Print("Maximum element in the Array: ")
	fmt.Println(max(arr))

	// MIN
	fmt.Printf("Minimum element in the Array: ")
	minElement := min(arr)
	fmt.Println(minElement)

	// SUM
	fmt.Print("Sum of the Array: ")
	fmt.Println(sum(arr))

	// AVERAGE
	fmt.Print("Average of the Array: ")
	fmt.Println(avg(arr))

	// Reverse Array
	fmt.Print("Original Array: ", arr, "\n")
	reverseArray := reverse(arr)
	fmt.Println("Reverse Array: ", reverseArray)

	// Copy Array
	src := []int{1, 2, 3, 4, 5, 6}
	des := make([]int, lenArr(src))
	copyArr(src, des)
	fmt.Println("source Array: ", src)
	fmt.Println("Destinition Array: ", des)

	// Even Numbers:
	fmt.Println("Even Elements in the Array: ")
	printEven(arr)

	// Odd Numbers:
	fmt.Println("Odd Elements in the Array: ")
	printOdd(arr)

	// Array Multiplication:
	array := []int{3, 6, 9}
	fmt.Print("Multiply each element of an array by 2: ")
	MultiplyArr(array)
	fmt.Println(array)
}

// Array Basics

// Array Length:
func lenArr(arr []int) int {
	count := 0
	for range arr {
		count++
	}
	return count
}

// Traversing array
func traverse(arr []int) {
	fmt.Println("Traversing the Array: ")

	for _, element := range arr {
		fmt.Println(element)
	}
}

// Insert Element:
func insArr(arr []int, index, element int) []int {
	return append(arr[:index], append([]int{element}, arr[index:]...)...)
}

// Delete Element:
func delArr(arr []int, index int) []int {
	return append(arr[:index], arr[index+1:]...)
}

// Update Element:
func updateArr(arr []int, indx, new int) []int {
	for index := range arr {
		if index == indx {
			arr[index] = new
		}
	}
	return arr
}

// Access Element:
func AccessElement(arr []int, key int) int {
	for index, element := range arr {
		if element == key {
			return index
		}
	}
	return -1
}

// Find Maximum:
func max(arr []int) int {
	max := arr[0]
	for _, element := range arr {
		if element > max {
			max = element
		}
	}

	return max
}

// Find Minimum:
func min(arr []int) int {
	min := arr[0]
	for _, element := range arr {
		if element < min {
			min = element
		}
	}
	return min
}

// Sum Elements: Sum all elements in an array.
func sum(arr []int) int {
	sum := 0
	for _, element := range arr {
		sum += element
	}
	return sum
}

// Average of Elements:
func avg(arr []int) int {
	sum := 0
	for _, element := range arr {
		sum += element
	}
	avg := sum / len(arr)
	return avg
}

// Reverse Array:
/*
func reverse(arr []int) []int {
	left := 0
	right := len(arr) - 1
	for left < right {
		temp := arr[left]
		arr[left] = arr[right]
		arr[right] = temp
		left++
		right--
	}
	return arr
}
*/

// Reverse Array:
func reverse(arr []int) []int {
	left := 0
	right := len(arr) - 1
	for left < right {
		arr[left], arr[right] = arr[right], arr[left]
		left++
		right--
	}
	return arr
}

// Copy Array: Copy all elements of one array to another
func copyArr(src []int, des []int) {
	if lenArr(src) != lenArr(des) {
		fmt.Println("Length of Source Array doesn't match with Destination Array.")
		return
	}

	for index, element := range src {
		des[index] = element
	}
}

// Even Numbers: Write a function to print all even numbers in an array
func printEven(arr []int) {
	for _, element := range arr {
		if element%2 == 0 {
			fmt.Println(element)
		}
	}
}

// Odd Numbers: Write a function to print all odd numbers in an array
func printOdd(arr []int) {
	for _, element := range arr {
		if element%2 != 0 {
			fmt.Println(element)
		}
	}
}

// Array Multiplication: Multiply each element of an array by 2
func MultiplyArr(arr []int) {
	for index := range arr {
		arr[index] = arr[index] * 2
	}
}
